scripts/o3de/tests/test_register.py

```diff
@@ -248,3 +251,158 @@ def find_ancestor_dir(target_file_name: pathlib.PurePath, start_path: pathlib.Pa
+class TestRegisterProject:
+    engine_path = pathlib.PurePath('o3de')
+    project_path = pathlib.PurePath('TestProject')
+
+    @pytest.mark.parametrize(
+        'test_engine_name, engine_version, engine_api_versions,'
+        'registered_gem_versions, project_engine_name, project_engine_version, '
+        'project_gems, project_compatible_engines, project_engine_api_dependencies,'
+        'gem_compatible_engines, gem_engine_api_dependencies,'
+        'force, expected_result', [
+            # passes when registering without version information
+            pytest.param('o3de1', None, None, { 'gem1':'' }, None, None, ['gem1'], None, None, None, None, False, 0),
+            pytest.param('o3de2', '', None, { 'gem1':'' }, None, None, ['gem1'], None, None, None, None, False, 0),
+            # fails when compatible_engines has no match
+            pytest.param('o3de3', '2.3.4', None, { 'gem1':'' }, None, None, ['gem1'], ['o3de3==1.2.3'], None, None, None,  False, 1),
+            pytest.param('o3de4', '1.2.3', None, { 'gem1':'' }, "", "", ['gem1'], ['o3de1'], None, None, None, False, 1),
+            # passes when forced
+            pytest.param('o3de5', '1.2.3', None, { 'gem1':'' }, None, None, ['gem1'], ['o3de1'], None, None, None, True, 0),
+            # passes when compatible_engines has match
+            pytest.param('o3de6', '0.0.0', None, { 'gem1':'' }, None, None, ['gem1'], ['o3de6'], None, None, None, False, 0),
+            pytest.param('o3de7', '1.2.3', None, { 'gem1':'' }, None, None, ['gem1'], ['o3de7>=1.2.3','o3de-sdk==2.3.4'], None, None, None, False, 0),
+            # fails when uses gem that is not known compatible with version 1.2.3 
+            pytest.param('o3de8', '1.2.3', None, { 'gem1':'' }, None, None, ['gem1'], ['o3de8'], None, ['o3de==2.3.4'], None, False, 1),
+            # passes when uses gem that is known compatible with version 1.2.3 
+            pytest.param('o3de9', '1.2.3', None, { 'gem1':'' }, None, None, ['gem1'], ['o3de9'], None, ['o3de9==1.2.3'], None, False, 0),
+            # passes when compatible engine not found but compatible api found
+            pytest.param('o3de10', '1.2.3', {'api':'1.2.3'}, { 'gem1':'' }, "", "", ['gem1'], ['o3de10==2.3.4'], ['api==1.2.3'], None, None, False, 0),
+            # passes when compatible engine found and no compatible api found
+            pytest.param('o3de11', '1.2.3', {'api':'2.3.4'}, { 'gem1':'' }, "", "", ['gem1'], ['o3de11==1.2.3'], ['api==1.2.3'], None, None, False, 0),
+            # fails when no compatible engine or api found
+            pytest.param('o3de12', '1.2.3', {'api':'2.3.4'}, { 'gem1':'' }, None, None, ['gem1'], ['o3de12==3.4.5'], ['api==4.5.6'], None, None, False, 1),
+            # passes when uses gem with compatible engine and api found
+            pytest.param('o3de13', '1.2.3', {'api':'2.3.4'}, { 'gem1':'' }, None, None, ['gem1'], None, None, ['o3de13>=1.0.0'], ['api~=1.2.3'] , False, 0),
```

Comment:
In this case the API is not compatible due to being a different version

Reply:
This is another situation where I wanted to let the developer allow a match in `compatible_engine` or `engine_api_dependencies` - this could be a fairly common case where a gem owner wants to say their gem is compatible with an engine version or some API version - if we require both then the developer would need to be sure to include all API versions for that API from every compatible_engine entry.  In the case above they would need to add a version specifier to include the newer API version like so:
```json
'api~=1.2.3',
'api==2.3.4'
```
or similar